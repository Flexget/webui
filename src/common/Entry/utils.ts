import { RawEntry, CardType, FieldNames, MappingType, GettersType } from './types';
import { Entry, DefaultEntry } from './fields';
import {
  RawMovieEntry,
  MovieEntry,
  movieFieldList,
  MovieFields,
  MovieFieldNames,
} from './fields/movies';
import {
  RawSeriesEntry,
  SeriesEntry,
  seriesFieldList,
  SeriesFields,
  SeriesFieldNames,
} from './fields/series';

import {
  RawEpisodeEntry,
  EpisodeEntry,
  EpisodeFields,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !!entry.seriesSeason;
const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<MovieFields>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<SeriesFields>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<EpisodeFields>(EpisodeFieldNames, episodesFieldList);

const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    type: CardType.Movie,
    ...getMovieProps(entry),
  };
};

const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  type: CardType.Series,
  ...getSeriesProps(entry),
});

const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  type: CardType.Episode,
  ...getEpisodeProps(entry),
});

const toDefaultEntry = (entry: RawEntry): DefaultEntry => ({
  ...entry,
  type: CardType.Default,
});

export const toEntry = (entry: RawEntry): Entry => {
  if (isMovie(entry)) {
    return toMovieEntry(entry);
  }
  if (isSeries(entry)) {
    return toSeriesEntry(entry);
  }
  if (isEpisode(entry)) {
    return toEpisodeEntry(entry);
  }
  return toDefaultEntry(entry);
};
