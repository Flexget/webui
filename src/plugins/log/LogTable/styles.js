import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { Table as RVTable } from 'react-virtualized';

const color = (theme, hex) => css`
  ${theme.palette.type === 'light'
    ? css`
        background-color: ${hex};
      `
    : css`
        color: ${hex};
      `}
`;

export const Table = styled(RVTable)`
  font-size: 1rem;

  & .error {
    ${({ theme }) => color(theme, '#f2dede')};
  }

  & .critical {
    ${({ theme }) => color(theme, '#f2dede')};
  }

  & .warning {
    ${({ theme }) => color(theme, '#fcf8e3')};
  }
`;
