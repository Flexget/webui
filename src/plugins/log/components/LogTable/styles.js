import styled from '@emotion/styled';
import { Table as RVTable } from 'react-virtualized';

export const Table = styled(RVTable)`
  font-size: 1rem;

  & .error {
    background-color: #f2dede;
  }

  & .critical {
    background-color: #f2dede;
  }

  & .warning {
    background-color: #fcf8e3;
  }
`;
