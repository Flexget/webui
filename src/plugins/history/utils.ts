import { groupBy } from 'utils/array';
import { HistoryItem, GroupByFields } from './types';

export const getGroupedHistory = (items: HistoryItem[], grouping: GroupByFields) =>
  groupBy(
    items,
    grouping,
    grouping === GroupByFields.Time ? (i: string) => i.split('T')[0] : undefined,
  );
