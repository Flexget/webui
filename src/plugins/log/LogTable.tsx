import React, { FC } from 'react';
import { AutoSizer, Column, Table } from 'react-virtualized';
import 'react-virtualized/styles.css';
import { css, ClassNames } from '@emotion/core';
import { Theme, useTheme } from '@material-ui/core';
import { LogMessage, LogLevel } from 'plugins/log/types';

const color = (theme: Theme, hex: string) =>
  theme.palette.type === 'light'
    ? css`
        background-color: ${hex};
      `
    : css`
        color: ${hex};
      `;

const table = css`
  font-size: 1rem;
`;

interface Props {
  messages: LogMessage[];
}

const LogTable: FC<Props> = ({ messages }) => {
  const theme = useTheme();

  return (
    <ClassNames>
      {({ css: cssString }) => {
        const cssMap = {
          [LogLevel.Error]: cssString(color(theme, '#f2dede')),
          [LogLevel.Critical]: cssString(color(theme, '#f2dede')),
          [LogLevel.Warning]: cssString(color(theme, '#fcf8e3')),
        };

        return (
          <AutoSizer>
            {({ height, width }) => {
              let actualWidth = width;
              let actualHeight = height;
              if (process.env.NODE_ENV === 'test') {
                actualWidth = window.innerWidth;
                actualHeight = window.innerHeight;
              }
              return (
                <Table
                  css={table}
                  rowCount={messages.length}
                  rowHeight={20}
                  headerHeight={50}
                  width={actualWidth}
                  height={actualHeight}
                  rowGetter={({ index }) => messages[index]}
                  rowClassName={({ index }) => cssMap[messages[index]?.logLevel]}
                >
                  <Column label="Time" dataKey="timestamp" width={100} />
                  <Column label="Level" dataKey="logLevel" width={100} />
                  <Column label="Plugin" dataKey="plugin" width={100} />
                  <Column label="Task" dataKey="task" width={100} />
                  <Column label="Message" dataKey="message" width={100} flexGrow={1} />
                </Table>
              );
            }}
          </AutoSizer>
        );
      }}
    </ClassNames>
  );
};

export default LogTable;
