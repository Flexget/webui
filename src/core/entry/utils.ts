import { RawEntry, FieldNames, MappingType, GettersType, BaseEntry } from './types';
import { MovieEntry, movieFieldList, RawMovieEntry, MovieFieldNames } from './fields/movies';
import { SeriesEntry, seriesFieldList, SeriesFieldNames, RawSeriesEntry } from './fields/series';
import {
  EpisodeEntry,
  RawEpisodeEntry,
  episodesFieldList,
  EpisodeFieldNames,
} from './fields/episodes';

export const isMovie = (entry: RawEntry): entry is RawMovieEntry => !!entry.movieName;
export const isSeries = (entry: RawEntry): entry is RawSeriesEntry =>
  !!entry.seriesName && !entry.seriesSeason;
export const isEpisode = (entry: RawEntry): entry is RawEpisodeEntry & SeriesEntry =>
  !!(entry.seriesSeason && entry.seriesEpisode && entry.seriesName);

const getPropFn = <U>(mapping: MappingType<U>) => {
  const mapByField = mapping.reduce<Partial<Record<FieldNames<U>, string[]>>>(
    (res, item) =>
      Object.entries(item).reduce(
        (out, [k, v]) => ({
          ...out,
          [k]: [...(out[k] ?? []), v],
        }),
        res,
      ),
    {},
  );
  return (entry: U, field: FieldNames<U>) =>
    (mapByField[field] ?? ([] as string[])).reduce((res, val) => res ?? entry[val], undefined);
};

const makeGetters = <T>(names: Record<string, FieldNames<T>>, mapping: MappingType<T>) => {
  const getProp = getPropFn<T>(mapping);
  return (entry: T) =>
    Object.values(names).reduce(
      (obj, key: FieldNames<T>) => ({
        ...obj,
        get [key]() {
          return getProp(entry, key);
        },
      }),
      {} as GettersType<T>,
    );
};

const getMovieProps = makeGetters<RawMovieEntry>(MovieFieldNames, movieFieldList);
const getSeriesProps = makeGetters<RawSeriesEntry>(SeriesFieldNames, seriesFieldList);
const getEpisodeProps = makeGetters<RawEpisodeEntry>(EpisodeFieldNames, episodesFieldList);

export const toMovieEntry = (entry: RawMovieEntry): MovieEntry => {
  return {
    ...entry,
    ...getMovieProps(entry),
  };
};

export const toSeriesEntry = (entry: RawSeriesEntry): SeriesEntry => ({
  ...entry,
  ...getSeriesProps(entry),
});

export const toEpisodeEntry = (entry: RawEpisodeEntry): EpisodeEntry => ({
  ...entry,
  ...getEpisodeProps(entry),
});

export const toDefaultEntry = (entry: BaseEntry): BaseEntry => ({
  ...entry,
});
