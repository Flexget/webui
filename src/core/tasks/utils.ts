import { RawEntry } from 'core/entry/types';

export const toExecuteRequest = (entry?: RawEntry) => {
  if (entry) {
    const { id, title, originalUrl, ...fields } = entry;
    return { title, url: originalUrl, fields };
  }
  return { title: '', url: '', fields: {} };
};
