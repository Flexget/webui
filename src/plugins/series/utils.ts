import { RawEntry } from 'core/entry/types';
import { Show, Episode } from './types';

export const showToEntry = (show: Show): RawEntry => ({
  title: show.name,
  originalUrl: '',
  quality: 'unknown',
  seriesName: show.name,
});

export const episodeToEntry = (show: Show, episode: Episode): RawEntry => ({
  title: episode.identifier,
  originalUrl: '',
  quality: 'unknown',
  seriesName: show.name,
  seriesSeason: episode.season,
  seriesEpisode: episode.number,
  seriesId: episode.identifier,
});
