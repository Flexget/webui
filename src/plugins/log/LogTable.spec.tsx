import React from 'react';
import { shallow } from 'enzyme';
import { themed } from 'utils/tests';
import LogTable from './LogTable';

describe('plugins/log/LogTable', () => {
  describe('LogTable', () => {
    it('renders correctly', () => {
      const wrapper = shallow(themed(<LogTable messages={[]} />));
      expect(wrapper).toMatchSnapshot();
    });
  });
});
