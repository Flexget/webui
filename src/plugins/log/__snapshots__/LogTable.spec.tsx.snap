// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/log/LogTable LogTable renders correctly 1`] = `
<Provider>
  <Theme>
    <LogTable
      messages={Array []}
    />
  </Theme>
</Provider>
`;
